cli/internal/cacheitem/restore_directory.go

```diff
@@ -0,0 +1,109 @@
+package cacheitem
+
+import (
+	"archive/tar"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/vercel/turborepo/cli/internal/turbopath"
+)
+
+// restoreDirectory restores a directory.
+func restoreDirectory(anchor turbopath.AbsoluteSystemPath, header *tar.Header, reader *tar.Reader) (turbopath.AnchoredSystemPath, error) {
+	processedName, err := canonicalizeName(header.Name)
+	if err != nil {
+		return "", err
+	}
+
+	// We need to traverse `processedName` from base to root split at
+	// `os.Separator` to make sure we don't end up following a symlink
+	// outside of the restore path.
+
+	// Create the directory.
+	if err := safeMkdirAll(anchor, processedName, header.Mode); err != nil {
+		return "", err
+	}
+
+	return processedName, nil
+}
+
+// safeMkdirAll creates all directories, assuming that the leaf node is a directory.
+// FIXME: Recheck the symlink cache before creating a directory.
+func safeMkdirAll(anchor turbopath.AbsoluteSystemPath, processedName turbopath.AnchoredSystemPath, mode int64) error {
+	// Iterate through path segments by os.Separator, appending them onto the anchor.
+	// Check to see if that path segment is a symlink with a target outside of anchor.
+	pathSegments := strings.Split(processedName.ToString(), string(os.PathSeparator))
+
+	calculatedAnchor := anchor
+	var checkPathErr error
+	for _, segment := range pathSegments {
+		calculatedAnchor, checkPathErr = checkPath(anchor, calculatedAnchor, turbopath.RelativeSystemPath(segment))
```

Comment:
Is this going to `lstat` every path segment on every file we restore?

Reply:
Not once I throw `LstatCachedFile` in here.

Reply:
Hmm. Maybe, but it looks like you'd need to cache the lstat output for every intermediate path we're restoring.

What's the consequence of not doing this check?

Reply:
The default version of "restore outside of tar root" is to change the header `name` field to `../../foo`. That's easy to check for and prevent (`checkName`). Much harder is when you create one symlink and one file: `external -> ../` and `external/foo`.

There is a fast path I can create here that would enable smarter behavior here (beyond just clever caching):
0. Confirm that the output directory is empty. (1 dirent)
1. Store all symlinks in their DAG and look through the resolved value there instead of hitting the OS. (rounds to zero)

The clever caching approach will take advantage of knowing that a (`turbo`-created) `tar.gz` is going to be breadth-first. I should only need to maintain a single cached resolution at any time. The number of lstat calls will be equivalent to the number of _folders_ in the archive.

The other option is to let people arbitrarily restore outside of their root if they use symlinks to do it.